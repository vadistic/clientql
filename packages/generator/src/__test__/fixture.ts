import path from 'path'
import { COMPLEX_TYPEDEFS, PRISMA_TYPEDEFS, VATS_TYPEDEFS } from '@clientql/testing'
import { getGeneratorProps } from '../generator'

export const prismaProps = getGeneratorProps(PRISMA_TYPEDEFS)

export const complexProps = getGeneratorProps(COMPLEX_TYPEDEFS)

export const vatsProps = getGeneratorProps(VATS_TYPEDEFS)

export const fileSnapPath = (filename: string, ...dir: string[]): [string, string] => [
  path.resolve(__dirname, '__file_snapshots__', ...dir, filename),
  filename,
]
