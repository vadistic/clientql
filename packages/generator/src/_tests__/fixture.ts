import {
  COMPLEX_TYPEDEFS,
  PRISMA_TYPEDEFS,
  VATS_TYPEDEFS,
} from '@graphql-clientgen/testing'
import path from 'path'
import { getGeneratorProps } from '../generator'

export const prismaProps = getGeneratorProps(PRISMA_TYPEDEFS)

export const complexProps = getGeneratorProps(COMPLEX_TYPEDEFS)

export const vatsProps = getGeneratorProps(VATS_TYPEDEFS)

export const fileSnapPath = (filename: string, ...dir: string[]) => [
  path.resolve(__dirname, '__file_snapshots__', ...dir, filename),
  filename,
]
