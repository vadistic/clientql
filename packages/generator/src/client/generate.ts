import {
  isInterfaceTypeDefinitonNode,
  isObjectTypeDefinitionNode,
  isUnionTypeDefinitionNode,
} from '@graphql-clientgen/core'
import { GeneratorProps } from '../generator'
import { traverseGraph } from '../traverse'
import { printClientInterfaces } from './client-interfaces'
import { printClientsResults } from './client-results'

/**
 * generate client lib
 *
 * moving parts:
 * - `client-typedefs` => filtered runtime typedefs
 * - `client-interfaces` => client proxy typings
 *     imports:
 *      - all client results
 *      - definitions: scalars & input types
 * - `client-definitions` => standard typescript codegen of all types
 * - `client-results` => possible operation results and their fragments
 *      - definitions: scalars
 * - `client-boilerplate` => typed client factory
 *      - root client interfaces
 */

export const generateClient = (props: GeneratorProps) => {
  const targets = resolvePossibleClients(props)

  const { fragmentsTs, resultsTs } = printClientsResults(props)(targets)

  const { clientsTs, rootClientsTs } = printClientInterfaces(props)(targets)

  return clientsTs + '\n\n' + resultsTs
}

/**
 * find possible results of the operations
 */

const resolvePossibleClients = (props: GeneratorProps) => {
  const targets = new Set<string>()

  traverseGraph(props)((vtx, stack) => {
    // no roots
    if (stack.length === 1) {
      return
    }

    if (
      isObjectTypeDefinitionNode(vtx.value) ||
      isInterfaceTypeDefinitonNode(vtx.value) ||
      isUnionTypeDefinitionNode(vtx.value)
    ) {
      targets.add(vtx.name)
    }
  })

  return targets
}
