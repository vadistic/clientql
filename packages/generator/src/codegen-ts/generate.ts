import {
  isEnumTypeDefinitionNode,
  isInputObjectTypeDefinitionNode,
  isScalarTypeDefinitionNode,
  unwrapDocument
} from '@graphql-clientgen/shared'
import {
  codegenEnum,
  codegenInputObject,
  codegenObjectToClient,
  codegenObjectToType,
  codegenScalar
} from '.'
import { GeneratorProps } from '../config/config'

export const generateTypescript = async (props: GeneratorProps) => {
  const definitions = unwrapDocument(props.doc)

  const objectTypes = Object.values(props.astMap.types).map(val => val.node)

  const objectTypesClientsTypescript = objectTypes
    .map(codegenObjectToClient(props))
    .join('\n\n')

  const objectTypesTypesTypescript = objectTypes
    .map(codegenObjectToType(props))
    .join('\n\n')

  const inputObjectTypesTypescript = definitions
    .filter(isInputObjectTypeDefinitionNode)
    .map(codegenInputObject(props))
    .join('\n\n')

  const scalarTypesTypescript = definitions
    .filter(isScalarTypeDefinitionNode)
    .map(codegenScalar(props))
    .join('\n\n')

  const enumTypesTypescript = definitions
    .filter(isEnumTypeDefinitionNode)
    .map(codegenEnum(props))
    .join('\n\n')

  let result = ''

  result += `/* SCALAR TYPES */ \n\n`
  // TODO: Detect if it's declared or something
  result += `export type ID = string` + '\n\n'
  result += scalarTypesTypescript + '\n\n'

  result += `/* CLIENT TYPES */ \n\n`
  result += objectTypesClientsTypescript + '\n\n'

  result += `/* OBJECT TYPES */ \n\n`
  result += objectTypesTypesTypescript + '\n\n'

  result += `/* INPUT OBJECT TYPES */ \n\n`
  result += inputObjectTypesTypescript + '\n\n'

  result += `/* ENUM TYPES */ \n\n`
  result += enumTypesTypescript + '\n\n'

  return result
}
