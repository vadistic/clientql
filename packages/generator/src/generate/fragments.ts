import {
  isObjectTypeDefinitionNode,
  isRootTypeDefinitionNode,
  nonNull
} from '@graphql-clientgen/shared'
import { DocumentNode, GraphQLObjectType, GraphQLSchema, Kind } from 'graphql'
import { GeneratorProps } from '../config/config'
import {
  FragmentType,
  objectTypeToFragment
} from '../fragment/object-to-fragment'

export const generateFragments = ({ ast, schema }: GeneratorProps) => {
  const objectTypes = ast.definitions
    .filter(isObjectTypeDefinitionNode)
    .filter(
      el => isObjectTypeDefinitionNode(el) && !isRootTypeDefinitionNode(el)
    )

  const flatFragments = objectTypes
    .map(objectTypeToFragment(schema, FragmentType.FLAT))
    .filter(nonNull)

  const deepFragments = objectTypes
    .map(objectTypeToFragment(schema, FragmentType.DEEP))
    .filter(nonNull)

  const defaultFragments = objectTypes
    .map(objectTypeToFragment(schema, FragmentType.DEFAULT))
    .filter(nonNull)

  const fragments = [flatFragments, defaultFragments, deepFragments].flat(1)

  return {
    fragments,
    flatFragments,
    deepFragments,
    defaultFragments
  }
}
