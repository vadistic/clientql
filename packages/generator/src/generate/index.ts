export * from './graphql-fragments'
export * from './graphql-typedefs'

export * from './typings-definitions'
export * from './typings-arguments'
export * from './typings-client'
