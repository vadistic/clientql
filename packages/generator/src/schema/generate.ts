import fs from 'fs-extra'
import { parse, print } from 'graphql'
import { getRemoteSchema } from './get-remote-schema'
import { mergeExtensions } from './merge-extensions'

/**
 * --endpoint cli flag flag
 */
export const remoteSchema = async (url: string, merge: boolean = true) => {
  const res = await getRemoteSchema(url)

  if (res.status === 'err') {
    throw Error(res.message)
  }

  if (!merge) {
    return {
      typeDefs: res.typeDefs,
      doc: parse(res.typeDefs)
    }
  }

  const mergedDoc = mergeExtensions(parse(res.typeDefs))

  return {
    typeDefs: print(mergedDoc),
    doc: mergedDoc
  }
}

/**
 * --schema cli flag flag
 */
export const localSchema = async (file: string, merge: boolean = true) => {
  const typeDefs = await fs.readFile(file, 'utf-8')
  const doc = parse(typeDefs)

  if (!merge) {
    return {
      typeDefs,
      doc
    }
  }

  const mergedDoc = mergeExtensions(parse(typeDefs))

  return {
    typeDefs: print(mergedDoc),
    doc: mergedDoc
  }
}
