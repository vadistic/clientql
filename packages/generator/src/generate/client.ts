import {
  isObjectTypeDefinitionNode,
  unwrapDocument
} from '@graphql-clientgen/shared'
import { codegenObjectMetaToClient } from '../client'
import { getObjectTypeDefinitionNodeMeta } from '../codegen'
import { GeneratorProps } from '../config'

export const generateClient = async (props: GeneratorProps) => {
  const definitions = unwrapDocument(props.ast)

  const objectTypesClientsTyping = definitions
    .filter(isObjectTypeDefinitionNode)
    .map(getObjectTypeDefinitionNodeMeta)
    .map(codegenObjectMetaToClient(props))
    .join('\n\n')

  return objectTypesClientsTyping
}
