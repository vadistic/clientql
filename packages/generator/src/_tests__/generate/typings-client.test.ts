import { generateTypingsClient } from '../../generate'
import { fixtureProps } from '../fixture'

describe('generate typings > client', () => {
  it('match snapshot', async () => {
    const result = await generateTypingsClient(fixtureProps)

    expect(result).toMatchSnapshot()
  })
})
