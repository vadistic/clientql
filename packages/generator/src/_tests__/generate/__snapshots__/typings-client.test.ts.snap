// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate typings > client match snapshot 1`] = `
"/*
 *
 * ROOT CLIENT
 *
 */

export interface IRootClient {
  node: (id: ID) => Promise<INode | null>
  post: (where: IPostWhereUniqueInput) => Promise<IPost | null> & IPostClient
  posts: (args: {
    after?: string | null
    before?: string | null
    first?: number | null
    last?: number | null
    orderBy?: PostOrderByInput | null
    skip?: number | null
    where?: IPostWhereInput | null
  }) => Promise<Array<IPost>> & IPostClient
  postsConnection: (args: {
    after?: string | null
    before?: string | null
    first?: number | null
    last?: number | null
    orderBy?: PostOrderByInput | null
    skip?: number | null
    where?: IPostWhereInput | null
  }) => Promise<IPostConnection> & IPostConnectionClient
  user: (where: IUserWhereUniqueInput) => Promise<IUser | null> & IUserClient
  users: (args: {
    after?: string | null
    before?: string | null
    first?: number | null
    last?: number | null
    orderBy?: UserOrderByInput | null
    skip?: number | null
    where?: IUserWhereInput | null
  }) => Promise<Array<IUser>> & IUserClient
  usersConnection: (args: {
    after?: string | null
    before?: string | null
    first?: number | null
    last?: number | null
    orderBy?: UserOrderByInput | null
    skip?: number | null
    where?: IUserWhereInput | null
  }) => Promise<IUserConnection> & IUserConnectionClient
  createPost: (data: IPostCreateInput) => Promise<IPost> & IPostClient
  createUser: (data: IUserCreateInput) => Promise<IUser> & IUserClient
  deleteManyPosts: (where?: IPostWhereInput | null) => Promise<IBatchPayload> & IBatchPayloadClient
  deleteManyUsers: (where?: IUserWhereInput | null) => Promise<IBatchPayload> & IBatchPayloadClient
  deletePost: (where: IPostWhereUniqueInput) => Promise<IPost | null> & IPostClient
  deleteUser: (where: IUserWhereUniqueInput) => Promise<IUser | null> & IUserClient
  updateManyPosts: (args: {
    data: IPostUpdateManyMutationInput
    where?: IPostWhereInput | null
  }) => Promise<IBatchPayload> & IBatchPayloadClient
  updateManyUsers: (args: {
    data: IUserUpdateManyMutationInput
    where?: IUserWhereInput | null
  }) => Promise<IBatchPayload> & IBatchPayloadClient
  updatePost: (args?: {
    data: IPostUpdateInput
    where: IPostWhereUniqueInput
  }) => Promise<IPost | null> & IPostClient
  updateUser: (args?: {
    data: IUserUpdateInput
    where: IUserWhereUniqueInput
  }) => Promise<IUser | null> & IUserClient
  upsertPost: (args: {
    create: IPostCreateInput
    update: IPostUpdateInput
    where: IPostWhereUniqueInput
  }) => Promise<IPost> & IPostClient
  upsertUser: (args: {
    create: IUserCreateInput
    update: IUserUpdateInput
    where: IUserWhereUniqueInput
  }) => Promise<IUser> & IUserClient
}

/*
 *
 * CLIENTS
 *
 */

export interface IAggregatePostClient extends IFragmentable {
  count: () => Promise<number>
}

export interface IAggregateUserClient extends IFragmentable {
  count: () => Promise<number>
}

export interface IBatchPayloadClient extends IFragmentable {
  count: () => Promise<number>
}

export interface IPageInfoClient extends IFragmentable {
  endCursor: () => Promise<string | null>
  hasNextPage: () => Promise<boolean>
  hasPreviousPage: () => Promise<boolean>
  startCursor: () => Promise<string | null>
}

export interface IPostClient extends IFragmentable {
  author: () => Promise<IUser> & IUserClient
  content: () => Promise<string | null>
  createdAt: () => Promise<DateTime>
  id: () => Promise<ID>
  published: () => Promise<boolean>
  title: () => Promise<string>
  updatedAt: () => Promise<DateTime>
}

export interface IPostConnectionClient extends IFragmentable {
  aggregate: () => Promise<IAggregatePost> & IAggregatePostClient
  edges: () => Promise<Array<IPostEdge>> & IPostEdgeClient
  pageInfo: () => Promise<IPageInfo> & IPageInfoClient
}

export interface IPostEdgeClient extends IFragmentable {
  cursor: () => Promise<string>
  node: () => Promise<IPost> & IPostClient
}

export interface IUserClient extends IFragmentable {
  email: () => Promise<string>
  id: () => Promise<ID>
  name: () => Promise<string | null>
  posts: (args?: {
    after?: string | null
    before?: string | null
    first?: number | null
    last?: number | null
    orderBy?: PostOrderByInput | null
    skip?: number | null
    where?: IPostWhereInput | null
  }) => Promise<Array<IPost> | null> & IPostClient
}

export interface IUserConnectionClient extends IFragmentable {
  aggregate: () => Promise<IAggregateUser> & IAggregateUserClient
  edges: () => Promise<Array<IUserEdge>> & IUserEdgeClient
  pageInfo: () => Promise<IPageInfo> & IPageInfoClient
}

export interface IUserEdgeClient extends IFragmentable {
  cursor: () => Promise<string>
  node: () => Promise<IUser> & IUserClient
}

"
`;
